modified-lar/src/main/scala/hmda/publication/lar/publication/ModifiedLarPublisher.scala

```diff
@@ -78,20 +98,74 @@ object ModifiedLarPublisher {
             bucket,
             s"$environment/modified-lar/$year/$fileName")
 
-          readRawData(submissionId)
-            .map(l => l.data)
-            .map(ByteString(_))
-            .via(framing("\n"))
-            .map(_.utf8String)
-            .map(_.trim)
-            .drop(1)
-            .map(s => ModifiedLarCsvParser(s).toCSV + "\n")
-            .map(s => ByteString(s))
-            .runWith(s3Sink)
+          def removeLei: Future[Int] =
+            modifiedLarRepo.deleteByLei(submissionId.lei)
+
+          val mlarSource: Source[ModifiedLoanApplicationRegister, NotUsed] =
+            readRawData(submissionId)
+              .map(l => l.data)
+              .drop(1)
+              .map(s => ModifiedLarCsvParser(s))
+
+          val s3Out: Sink[ModifiedLoanApplicationRegister,
+            Future[MultipartUploadResult]] =
+            Flow[ModifiedLoanApplicationRegister]
+              .map(mlar => mlar.toCSV + "\n")
+              .map(ByteString(_))
+              .toMat(s3Sink)(Keep.right)
+
+          def postgresOut(parallelism: Int)
+          : Sink[ModifiedLoanApplicationRegister, Future[Done]] =
+            Flow[ModifiedLoanApplicationRegister]
+              .map(
+                mlar =>
+                  EnrichedModifiedLoanApplicationRegister(
+                    mlar,
+                    indexTractMap.getOrElse(mlar.tract, Census())
+                  )
+              )
+              .mapAsync(parallelism)(enriched =>
+                modifiedLarRepo.insert(enriched, submissionId.toString))
+              .toMat(Sink.ignore)(Keep.right)
+
+          val graph = mlarSource
+            .alsoToMat(postgresOut(2))(Keep.right) // TODO: provide a way to make this configurable
+            .toMat(s3Out)(Keep.both)
+            .mapMaterializedValue {
+              // We listen on the completion of both materialized values but we only keep the S3 as the result
+              // since that is a meaningful value
+              case (futPostgresRes, futS3Res) =>
+                for {
+                  _ <- futPostgresRes
+                  s3Res <- futS3Res
+                } yield s3Res
+            }
+
+          val finalResult: Future[Unit] = for {
```

Comment:
It seems from the logic that removing old psql records and populating the new ones might happen simultaneously and could lead to a race condition. It seems like we should remove LEI and repopulate with new records synchronously.

Reply:
`removeLei ` and `postgresOut ` are declared as `def` as opposed to `val` so they will happen synchronously.